modules/lang-painless/src/main/java/org/elasticsearch/painless/LambdaBootstrap.java

```diff
@@ -0,0 +1,530 @@
+        }
+
+        iface.returnValue();
+        iface.endMethod();
+    }
+
+    /**
+     * Closes the {@link ClassWriter}.
+     */
+    private static void endLambdaClass(ClassWriter cw) {
+        cw.visitEnd();
+    }
+
+    /**
+     * Defines the {@link Class} for the lambda class using the same {@link Loader}
+     * that originally defined the class for the Painless script.
+     */
+    private static Class<?> createLambdaClass(
+            Loader loader,
+            ClassWriter cw,
+            String lambdaClassName) {
+
+        byte[] classBytes = cw.toByteArray();
+        return AccessController.doPrivileged((PrivilegedAction<Class<?>>)() ->
+            loader.defineLambda(lambdaClassName.replace('/', '.'), classBytes));
+    }
+
+    /**
+     * Creates an {@link ConstantCallSite} that will return the same instance
+     * of the generated lambda class every time this linked factory method is called.
+     */
+    private static CallSite createNoCaptureCallSite(
+            MethodType factoryMethodType,
+            Class<?> lambdaClass) {
+
+        Constructor<?> constructor = AccessController.doPrivileged(
+            (PrivilegedAction<Constructor<?>>)() -> {
+                try {
+                    return lambdaClass.getConstructor();
+                } catch (NoSuchMethodException nsme) {
+                    throw new IllegalStateException("unable to create lambda class", nsme);
+                }
+            });
+
+        try {
+            return new ConstantCallSite(MethodHandles.constant(
+                factoryMethodType.returnType(), constructor.newInstance()));
+        } catch (InstantiationException |
+            IllegalAccessException |
+            InvocationTargetException exception) {
+            throw new IllegalStateException("unable to create lambda class", exception);
+        }
+    }
+
+    /**
+     * Creates an {@link ConstantCallSite}
+     */
+    private static CallSite createCaptureCallSite(
+            Lookup lookup,
+            String factoryMethodName,
+            MethodType factoryMethodType,
+            Class<?> lambdaClass) {
+
+        try {
+            return new ConstantCallSite(
+                lookup.findStatic(lambdaClass, factoryMethodName, factoryMethodType));
+        } catch (NoSuchMethodException | IllegalAccessException exception) {
+            throw new IllegalStateException("unable to create lambda factory class", exception);
+        }
+    }
+
+    /**
+     * Links the delegate method to the returned {@link CallSite}.  The linked
+     * delegate method will use converted types from the interface method.  Using
+     * invokedynamic to make the delegate method call allows
+     * {@link MethodHandle#asType} to be used to do the type conversion instead
+     * of either a lot more code or requiring many {@link Definition.Type}s to be looked
+     * up at link-time.
+     */
+    public static CallSite delegateBootstrap(Lookup lookup,
```

Comment:
Going to leave this one as the methods are in order of which ones will be called first.